codegenerator/cli/templates/static/codegen/src/db/DbFunctionsImplementation.js

```diff
@@ -231,26 +231,45 @@ module.exports.deleteStaleEndOfBlockRangeScannedDataForChain = (
     ;`;
 };
 
-module.exports.recoverRegisteredDynamicContracts = (
+module.exports.deleteInvalidDynamicContractsOnRestart = (
   sql,
   chainId,
-  blockNumber
-) => sql`
-  SELECT *
-  FROM "public"."dynamic_contract_registry"
-  WHERE registering_event_block_number <= ${blockNumber} 
-  AND chain_id = ${chainId};`;
+  restartBlockNumber,
+  restartLogIndex
+) => {
+  return sql`
+    DELETE
+    FROM "public"."dynamic_contract_registry"
+    WHERE chain_id = ${chainId}
+    AND is_pre_registered = false
+    AND (
+      registering_event_block_number > ${restartBlockNumber} 
+      OR registering_event_block_number = ${restartBlockNumber}
+      AND registering_event_log_index > ${restartLogIndex}
+    );`;
+};
 
-module.exports.recoverPreRegisteredAndRegisteredDynamicContracts = (
+module.exports.deleteInvalidDynamicContractsHistoryOnRestart = (
   sql,
   chainId,
-  blockNumber
-) => sql`
+  restartBlockNumber,
+  restartLogIndex
+) => {
+  return sql`
+    DELETE
+    FROM "public"."dynamic_contract_registry_history"
+    WHERE entity_history_chain_id = ${chainId}
+    AND (
+      entity_history_block_number > ${restartBlockNumber} 
+      OR entity_history_block_number = ${restartBlockNumber}
+      AND entity_history_log_index > ${restartLogIndex}
+    );`;
```

Comment:
Cool makes sense.

Reply:
I guess the other option is that instead of deleting them on restart, we upsert them, but agree that deleting makes things consistent between restart runs and runs without restarts and removes potential for strange behavior on edge cases. :+1: